engine/src/main/java/org/pentaho/di/repository/kdr/KettleDatabaseRepositoryCreationHelper.java

```diff
@@ -0,0 +1,3131 @@
+      KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_ID_DATABASE_CONTYPE, KEY,
+      0 ) );
+    table.addValueMeta( new ValueMetaString(
+      KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_CODE,
+      KettleDatabaseRepository.REP_STRING_CODE_LENGTH, 0 ) );
+    table.addValueMeta( new ValueMetaString(
+      KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_DESCRIPTION,
+      KettleDatabaseRepository.REP_STRING_CODE_LENGTH, 0 ) );
+    sql =
+      database.getDDL(
+        schemaTable, table, null, false, KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_ID_DATABASE_CONTYPE,
+        false );
+
+    if ( !Utils.isEmpty( sql ) ) {
+      statements.add( sql );
+      if ( !dryrun ) {
+        if ( log.isDetailed() ) {
+          log.logDetailed( "executing SQL statements: " + Const.CR + sql );
+        }
+        database.execStatements( sql );
+        if ( log.isDetailed() ) {
+          log.logDetailed( "Created or altered table " + schemaTable );
+        }
+      }
+    } else {
+      if ( log.isDetailed() ) {
+        log.logDetailed( "Table " + schemaTable + " is OK." );
+      }
+    }
+
+    // If it's creating the table, go ahead and populate below...
+    //
+    ok_database_contype = sql.toUpperCase().contains( "CREATE TABLE" );
+    if ( ok_database_contype ) {
+      //
+      // Populate with data...
+      //
+      code = DatabaseMeta.dbAccessTypeCode;
+      desc = DatabaseMeta.dbAccessTypeDesc;
+
+      if ( !dryrun ) {
+        database.prepareInsert( table, null, tablename );
+      }
+
+      for ( int i = 0; i < code.length; i++ ) {
+        RowMetaAndData lookup = null;
+        if ( upgrade ) {
+          lookup =
+            database.getOneRow( "SELECT "
+              + repository.quote( KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_ID_DATABASE_CONTYPE )
+              + " FROM " + schemaTable + " WHERE "
+              + repository.quote( KettleDatabaseRepository.FIELD_DATABASE_CONTYPE_CODE ) + " = '" + code[i]
+              + "'" );
+        }
+        if ( lookup == null ) {
+          ObjectId nextid = new LongObjectId( i + 1 );
+          if ( !create ) {
+            nextid = repository.connectionDelegate.getNextDatabaseConnectionTypeID();
+          }
+
+          Object[] tableData = new Object[] { new LongObjectId( nextid ).longValue(), code[i], desc[i], };
+          if ( dryrun ) {
+            sql = database.getSQLOutput( null, tablename, table, tableData, null );
+            statements.add( sql );
+          } else {
+            database.setValuesInsert( table, tableData );
+            database.insertRow();
+          }
+        }
+      }
+
+      try {
+        if ( !dryrun ) {
+          database.closeInsert();
+        }
+        if ( log.isDetailed() ) {
+          log.logDetailed( "Populated table " + schemaTable );
+        }
+      } catch ( KettleException dbe ) {
+        throw new KettleException( "Unable to close insert after populating table " + schemaTable, dbe );
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 379:388, 1963:1972, 2921:2930.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.